docs/designs/master_redesign.md

```diff
@@ -0,0 +1,136 @@
+# Master Design
+
+This document deseribes an ElasticDL master design.
+
+## Motivation
+
+We plan to provide elastic training feature to customized AllReduce training codes and other deep learning frameworks. Currently the master is designed for ElasticDL's own training framework. This requires us to re-design the master.
+After this re-design, the master consists of a few modular components and only launches those required components when providing elastic training feature to others.
+
+## Master Components
+The master has four modular components: elastic pod manager, task manager, rendezvous server and ElasticDL training/evaluation service.
+
+### Elastic Pod Manager
+The elastic pod manager is responsible for:
+
+1. Create PS/worker pods;
+2. Monitor pod events, keep track all pods status;
+3. Relaunch failed worker pods when needed.
+
+### Task Manager
+
+The task manager is responsible for creating tasks. Workers will ask the task manager for new tasks and report task completion status to the task manager when needed. Optionally, the task mamager supports task fault-tolerance by recovering un-completed tasks from failed workers and re-assigning to other workers.
```

Comment:
The task manager is responsible for creating tasks and dispatching tasks to workers.